### ISSUE TITLE
VPN connection often disconnects and connects again

### ISSUE THREAD

**Issue by khenderick** [droscy/esphome#7]
First; I had the impression that you prefer reports here, to keep the upstream PR clean. If not, let me know so I can move the report.

It looks like my VPN connection is unstable; disconnecting every ~2 minutes for ~10 seconds. While that happens, I can have the logs to the remote device open without any visible disconnects (it however does log the status changes), but that might be expected since it's only disconnected for ~10 seconds.

A snippet from the logs:

```
...
[07:44:35][D][text_sensor:064]: 'Air quality': Sending state 'Good'
[07:44:44][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-09 07:41:21 CEST)
[07:44:52][D][sensor:094]: 'Wifi signal': Sending state -54.00000 dBm with 0 decimals of accuracy
[07:44:54][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-09 07:41:21 CEST)
[07:44:54][D][sensor:094]: 'Brightness': Sending state 3.13800 V with 2 decimals of accuracy
[07:45:01][D][sensor:094]: 'Latest Handshake': Sending state 1686289280.00000  with 0 decimals of accuracy
[07:45:04][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-09 07:41:21 CEST)
[07:45:06][D][scd30:181]: Got CO2=409.98ppm temperature=22.03°C humidity=54.02%
[07:45:06][D][sensor:094]: 'CO2': Sending state 409.97522 ppm with 1 decimals of accuracy
[07:45:06][D][switch:012]: 'fan' Turning ON.
[07:45:06][D][light:035]: 'light_green' Setting:
[07:45:06][D][light:050]:   Brightness: 60%
[07:45:06][D][light:084]:   Transition length: 1.0s
[07:45:06][D][light:035]: 'light_orange' Setting:
[07:45:06][D][light:084]:   Transition length: 1.0s
[07:45:06][D][light:035]: 'light_red' Setting:
[07:45:06][D][light:084]:   Transition length: 1.0s
[07:45:06][D][text_sensor:064]: 'Air quality': Sending state 'Good'
[07:45:06][D][sensor:094]: 'Temperature': Sending state 22.02792 °C with 2 decimals of accuracy
[07:45:06][D][sensor:094]: 'Humidity': Sending state 54.01764 % with 2 decimals of accuracy
[07:45:14][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-09 07:41:21 CEST)
[07:45:22][D][binary_sensor:036]: 'VPN Status': Sending state OFF
[07:45:22][D][sensor:094]: 'Wifi signal': Sending state -54.00000 dBm with 0 decimals of accuracy
[07:45:24][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-09 07:43:22 CEST)
[07:45:32][D][binary_sensor:036]: 'VPN Status': Sending state ON
[07:45:34][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-09 07:43:22 CEST)
[07:45:35][D][pm1006:091]: Got PM2.5 Concentration: 32 µg/m³
[07:45:35][D][sensor:094]: 'PM': Sending state 32.00000 µg/m³ with 0 decimals of accuracy
...
```

The connection status history:

![image](https://github.com/droscy/esphome/assets/1453214/0cf989f9-c501-4808-b333-b02b0999fb55)

My configuration (or at least the relevant part from what I understand)

device.yaml:
```yaml
substitutions:
  static_ip: "192.x.10.53"
  vpn_ip: "10.y.25.14"
  vpn_private_key: !secret wireguard_device

packages:
  ...
  vpn: !include includes/common/vpn.yaml
```

vpn.yaml:
```yaml
external_components:
  - source: github://pr#4256
    components:
      - wireguard
      - wireguard_status
      - wireguard_handshake

time:
  - platform: sntp
    id: sntp_time

wireguard:
  id: vpn
  address: "${vpn_ip}"
  private_key: "${vpn_private_key}"
  peer_public_key: !secret wireguard_server
  peer_endpoint: "<domain_name>"
  peer_persistent_keepalive: 20
  peer_allowed_ips:
    - "10.y.0.0/16"

binary_sensor:
  - platform: wireguard_status
    name: "VPN Status"
    update_interval: 10s

sensor:
  - platform: wireguard_handshake
    name: "Latest Handshake"
    update_interval: 60s
```

**Comment by droscy** [comment]
> I had the impression that you prefer reports here, to keep the upstream PR clean.

Yes, I do, thanks.

> It looks like my VPN connection is unstable; disconnecting every ~2 minutes for ~10 seconds.

I'm not an expert of wireguard protocol itself, but the timing you reported matches with `REKEY_AFTER_TIME` that is 120 seconds and the 10s delay is the refresh of `wireguard_status` binary sensor.

From your log you have

```
[07:45:14][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-09 07:41:21 CEST)
[07:45:22][D][binary_sensor:036]: 'VPN Status': Sending state OFF
[07:45:24][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-09 07:43:22 CEST)
[07:45:32][D][binary_sensor:036]: 'VPN Status': Sending state ON
```

the wireguard component never catch the "disconnection" while it is reported by the sensor (and then to Home Assistant). It happened between 07:45:14 and 07:45:22 and only 2 seconds later everything was connected again.

It looks like spurious reporting but the the sensor rely on the same function of main component, so it is very strange.

Please @khenderick test again adding `update_interval: 2s` under `wireguard`, so we have a log from main component every 2 seconds and we see if, and when, it will catch the disconnection.

**Comment by khenderick** [comment]
An extract of the logs with this setting added;

```
...
[14:28:54][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 14:26:15 CEST)
... (message repeats every 2s) ...
[14:29:12][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 14:26:15 CEST)
[14:29:14][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 14:26:15 CEST)
[14:29:16][W][wireguard:075]: WireGuard remote peer is offline (latest handshake timestamp not available)
[14:29:18][D][wireguard:078]: WireGuard remote peer is offline (latest handshake timestamp not available)
[14:29:20][D][wireguard:078]: WireGuard remote peer is offline (latest handshake timestamp not available)
[14:29:22][D][wireguard:078]: WireGuard remote peer is offline (latest handshake timestamp not available)
[14:29:23][D][binary_sensor:036]: 'VPN Status': Sending state OFF
[14:29:24][D][wireguard:078]: WireGuard remote peer is offline (latest handshake timestamp not available)
[14:29:26][D][wireguard:078]: WireGuard remote peer is offline (latest handshake timestamp not available)
... (message repeats every 2s) ...
[14:31:12][D][wireguard:078]: WireGuard remote peer is offline (latest handshake timestamp not available)
[14:31:14][D][wireguard:078]: WireGuard remote peer is offline (latest handshake timestamp not available)
[14:31:16][I][wireguard:068]: WireGuard remote peer is online (latest handshake 2023-06-10 14:29:15 CEST)
[14:31:18][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 14:29:15 CEST)
[14:31:20][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 14:29:15 CEST)
[14:31:22][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 14:29:15 CEST)
[14:31:23][D][binary_sensor:036]: 'VPN Status': Sending state ON
[14:31:24][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 14:29:15 CEST)
[14:31:26][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 14:29:15 CEST)
[14:31:28][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 14:29:15 CEST)
...
```

Important to note here:
* I've only kept wireguard-related lines; a bunch of other components were just printing their usual logs without anything special
* As stated in the logs, I've left out identical messages that were repeated at an exact 2s interval without any missing entries
* I've watched these logs appear at the correct timing, using the VPN (the device is at a remote location and I cannot access the device locally) via the ESPHome addon view logs functionality. In other words, there was no pause in the messages or anything that would indicate the VPN connection was actually disconnected at some point.

I also want to thank you for this awesome work in making wireguard available on ESPHome.

**Comment by khenderick** [comment]
Another interesting log snippet:

```
...
[14:35:04][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 14:31:16 CEST)
[14:35:06][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 14:31:16 CEST)
[14:35:08][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 14:31:16 CEST)
[14:35:10][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 14:31:16 CEST)
[14:35:12][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 14:31:16 CEST)
[14:35:14][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 14:31:16 CEST)
[14:35:16][W][wireguard:075]: WireGuard remote peer is offline (latest handshake 2023-06-10 14:31:16 CEST)
[14:35:18][I][wireguard:068]: WireGuard remote peer is online (latest handshake 2023-06-10 14:33:16 CEST)
[14:35:20][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 14:33:16 CEST)
[14:35:22][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 14:33:16 CEST)
...
```

**Comment by droscy** [comment]
> I also want to thank you for this awesome work in making wireguard available on ESPHome.

Happy to contribute ;-)

> I've watched these logs appear at the correct timing, using the VPN (the device is at a remote location and I cannot access the device locally) via the ESPHome addon view logs functionality. In other words, there was no pause in the messages or anything that would indicate the VPN connection was actually disconnected at some point.

Oh, wow, I didn't understand that. So the log reports "disconnection" but you are logging remotely with no disconnection at all, so it is a wrong reporting


```
[14:29:14][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 14:26:15 CEST)
[14:29:16][W][wireguard:075]: WireGuard remote peer is offline (latest handshake timestamp not available)
```

This above kind of logs should not happen because the lastest handshake should never reset, maybe this is a bug, I'll check.

```
[14:35:14][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 14:31:16 CEST)
[14:35:16][W][wireguard:075]: WireGuard remote peer is offline (latest handshake 2023-06-10 14:31:16 CEST)
[14:35:18][I][wireguard:068]: WireGuard remote peer is online (latest handshake 2023-06-10 14:33:16 CEST)
```

This above instead is logically correct because the connection will be reported as offline if the latest handshake is more than "two times `REKEY_AFTER_TIME`" in the past (so 4 minutes in the past): latest handshake is `14:31:16` and at 14:35:16 (4 minutes later) the component reports offline. Then 2 seconds later the latest handshake is updated `14:33:16` and it is in the past less then 4 minutes so the device is online. **But**, if the handshake, in your setup, has refreshed less often than 120 seconds, I would expect the new value to be almost the same of current time instead at 14:35:18 it became `14:33:16`, 2 minutes earlier. That's strange too. I'll check.

I'll try to replicate your setup. What is the OS on the remote side? How did you install wireguard on the remote side?

In the mean time, could you comment out lines from 123 to 135 of `wireguard.cpp` (https://github.com/lhoracek/esphome/blob/feature/wireguard_component/esphome/components/wireguard/wireguard.cpp) and test again?

**Comment by khenderick** [comment]
The remote side runs `Raspbian GNU/Linux 11 (bullseye)` with `wireguard/stable,now 1.0.20210223-1 all` installed.

The relevant snippet from `wg0.conf`:

```
[Interface]
PrivateKey = xxx
Address = 10.x.25.1/24
ListenPort = 51820

[Peer]
PublicKey = xxx
AllowedIPs = 10.x.25.14/32
```

I'll make the code changes and add another comment with the results.

**Comment by khenderick** [comment]
When running the suggested changes:

```
...
[19:51:20][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 17:48:24 GMT)
[19:51:22][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 17:48:24 GMT)
[19:51:25][W][wireguard:075]: WireGuard remote peer is offline (latest handshake 2023-06-10 17:48:24 GMT)
WARNING 10.4.25.14: Connection error occurred: EOF received
INFO Processing unexpected disconnect from ESPHome API for 10.4.25.14
WARNING Disconnected from API
INFO Successfully connected to 10.4.25.14
[19:51:26][I][wireguard:068]: WireGuard remote peer is online (latest handshake timestamp not available)
[19:51:28][D][wireguard:071]: WireGuard remote peer is online (latest handshake timestamp not available)
...
[19:53:22][D][wireguard:071]: WireGuard remote peer is online (latest handshake timestamp not available)
[19:53:24][D][wireguard:071]: WireGuard remote peer is online (latest handshake timestamp not available)
[19:53:26][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 17:51:25 GMT)
[19:53:28][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 17:51:25 GMT)
...
[19:55:22][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 17:51:25 GMT)
[19:55:24][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 17:51:25 GMT)
[19:55:26][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 17:53:26 GMT)
[19:55:28][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 17:53:26 GMT)
...
[19:56:48][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 17:53:26 GMT)
[19:56:50][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 17:53:26 GMT)
[19:56:51][D][sntp:078]: Synchronized time: 2023-06-10 17:56:51
[19:56:51][D][wireguard:174]: WireGuard connection already started
[19:56:52][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 17:53:26 GMT)
[19:56:54][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 17:53:26 GMT)
...
[19:57:24][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 17:53:26 GMT)
[19:57:26][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 17:53:26 GMT)
[19:57:28][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 17:55:26 GMT)
[19:57:30][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 17:55:26 GMT)
...
```

It seems like it's always seeing the previous handshake, and not the actual latest one. It also looks like it's not reporting offline anymore. I'll keep an eye on it.

**Comment by khenderick** [comment]
Another thing I'd like to add; it seems with the 2s update interval, it switched status a lot more (to be clear, that's _before_ commenting out lines 123 -> 135)

![image](https://github.com/droscy/esphome/assets/1453214/14624259-9aeb-41eb-b174-b4eb1d99a45d)

Update, it seems commenting out lines 123 -> 135 do help:

![image](https://github.com/droscy/esphome/assets/1453214/858d6bca-69de-40e4-8b4c-eaacd7c77c75)

**Comment by droscy** [comment]
Hello @khenderick, I have random disconnections sometimes but, honestly, I didn't care about them too much. Observing them, they have the same behaviour you reported.

These are the relevant parts (I should have logged also by serial port, in order to compare remote and local logs, however): 

```bash
# a network error (maybe in wifi, I don't know) that resets the handshake
[22:22:18][D][sensor:094]: 'Wireguard Latest Handshake': Sending state 1686428416.00000  with 0 decimals of accuracy
[22:22:19][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 22:19:25 CEST)
[22:22:21][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 22:19:25 CEST)
[22:22:23][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 22:19:25 CEST)
WARNING 172.16.34.100: Connection error occurred: [Errno 104] Connection reset by peer
INFO Processing unexpected disconnect from ESPHome API for 172.16.34.100
WARNING Disconnected from API
INFO Successfully connected to 172.16.34.100
[22:22:29][D][wireguard:078]: WireGuard remote peer is offline (latest handshake timestamp not available)
[22:22:31][D][wireguard:078]: WireGuard remote peer is offline (latest handshake timestamp not available)
[22:22:31][D][binary_sensor:036]: 'Wireguard Status': Sending state OFF
[22:22:33][D][wireguard:078]: WireGuard remote peer is offline (latest handshake timestamp not available)
[22:22:35][D][wireguard:078]: WireGuard remote peer is offline (latest handshake timestamp not available)
[22:22:36][V][sensor:043]: 'Wireguard Test WiFi Signal': Received new state -70.000000
[22:22:36][D][sensor:094]: 'Wireguard Test WiFi Signal': Sending state -70.00000 dBm with 0 decimals of accuracy
[22:22:37][D][wireguard:078]: WireGuard remote peer is offline (latest handshake timestamp not available)
[...]
[22:24:27][D][wireguard:078]: WireGuard remote peer is offline (latest handshake timestamp not available)
[22:24:28][V][component:204]: Component api took a long time for an operation (0.13 s).
[22:24:28][V][component:205]: Components should block for at most 20-30ms.
[22:24:29][I][wireguard:068]: WireGuard remote peer is online (latest handshake 2023-06-10 22:22:27 CEST)

# remote peer reported as offline but really online with handshake 2 minutes behind
[22:26:25][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 22:22:27 CEST)
[22:26:27][W][wireguard:075]: WireGuard remote peer is offline (latest handshake 2023-06-10 22:22:27 CEST)
[22:26:29][D][wireguard:078]: WireGuard remote peer is offline (latest handshake 2023-06-10 22:22:27 CEST)
[22:26:31][I][wireguard:068]: WireGuard remote peer is online (latest handshake 2023-06-10 22:24:28 CEST)
[22:26:33][D][wireguard:071]: WireGuard remote peer is online (latest handshake 2023-06-10 22:24:28 CEST)
```

To get the status of remote peer I added the check on latest handshake delay but probably it is only a source of noise, not a quickly reporting of peer being offline. People that developed the underlying wireguard library have much more knowledge of it than me and I should have trusted them more :-(

And the latest handshake, directly retrieved from "current key" timestamp, can be `0` in case of errors or, probably, due to the key rotation, so I should report only the latest *valid* handshake.

**Comment by droscy** [comment]
Moving the sensors inside wireguard component I have handled the case of lasted handshake being `0` and removed my noisy check.

Could you test from my branch `fix/wg/disconnection/1`?

**Important**: that branch has breaking changes, you need to use the following configuration!

```yaml
external_components:
  - source:
      type: git
      url: https://github.com/droscy/esphome
      ref: fix/wg/disconnection/1
    components:
      - wireguard  # only wireguard, not the others

wireguard:
  # configure as usual
  #update_interval: 2s  # if you want

binary_sensor:
  - platform: wireguard
    status:
      name: 'WireGuard Status'

sensor:
  - platform: wireguard
    latest_handshake:
      name: 'WireGuard Latest Handshake'
```

**Comment by khenderick** [comment]
I think I made a mistake and use the wrong secrets file, so it has the incorrect keys. Oops :sweat_smile:. I'll try to get access to the device and give you feedback as soon as possible.

**Comment by khenderick** [comment]
We're back :+1:. Always nice to have a second device on that same remote network that I could use to port forward to ESPHome's webinterface to manually upload a new binary :sweat_smile: .

In any case, things seems to be working nice now:

```
...
[08:20:02][D][wireguard:072]: WireGuard remote peer is online (latest handshake 2023-06-14 08:17:57 CEST)
[08:20:04][D][wireguard:072]: WireGuard remote peer is online (latest handshake 2023-06-14 08:17:57 CEST)
[08:20:06][D][wireguard:072]: WireGuard remote peer is online (latest handshake 2023-06-14 08:17:57 CEST)
[08:20:08][D][wireguard:072]: WireGuard remote peer is online (latest handshake 2023-06-14 08:17:57 CEST)
[08:20:10][D][wireguard:072]: WireGuard remote peer is online (latest handshake 2023-06-14 08:18:07 CEST)
[08:20:10][D][sensor:094]: 'Latest Handshake': Sending state 1686723456.00000  with 0 decimals of accuracy
[08:20:12][D][wireguard:072]: WireGuard remote peer is online (latest handshake 2023-06-14 08:18:07 CEST)
[08:20:14][D][wireguard:072]: WireGuard remote peer is online (latest handshake 2023-06-14 08:18:07 CEST)
[08:20:16][D][wireguard:072]: WireGuard remote peer is online (latest handshake 2023-06-14 08:18:07 CEST)
[08:20:18][D][wireguard:072]: WireGuard remote peer is online (latest handshake 2023-06-14 08:18:07 CEST)
...
```

I also have a disconnect every now and then, but it's irregular, and maybe only once or twice a day. I consider these just normal network behavior and that's perfectly fine. They also show up in all sensors being unavailable, so the VPN actually being down for a brief moment.

**Comment by droscy** [comment]
> In any case, things seems to be working nice now:

Good.

> I also have a disconnect every now and then, but it's irregular, and maybe only once or twice a day. I consider these just normal network behavior and that's perfectly fine. They also show up in all sensors being unavailable, so the VPN actually being down for a brief moment.

I have them, too. Mostly when wifi is heavily under load. Maybe some packets get dropped...

If it is ok for you I will consider this issue closed as soon as I'll merge the code.

**Comment by khenderick** [comment]
It now works great as expected, thanks! You can indeed consider this issue closed :+1: .
